internal/subscriber/subscribe.go

```diff
@@ -74,66 +74,57 @@ func (s *SubscriberService) Start(ctx context.Context) error {
 	return nil
 }
 
-func nackLog(msg *pubsub.Message, err error) {
-	log.Printf("Subscriber error: %v\n", err)
-	msg.Nack() // TODO: Improve error handling by adding proper logging
-}
-
-// consumeMessage is a callback function for pubsub.Receive() which performs
-// updates to database according to incoming metadata.
+// processMessage handles incoming metadata and performs database operations based on the eventType of incoming *pubsub.Message
 //
-// Messages are expected to be unordered so the handling of incoming metadata has to
+// Messages are expected to be unordered. The handling of incoming metadata has to
 // be based on its update time and gracefully Nack()'d when necessary
-func (s *SubscriberService) consumeMessage(ctx context.Context, msg *pubsub.Message) {
+func processMessage(s *SubscriberService, msg *pubsub.Message) (ack bool, err error) {
+	// parse payload
 	var p payload
-	if err := json.Unmarshal(msg.Data, &p); err != nil {
-		nackLog(msg, err)
-		return
+	if err = json.Unmarshal(msg.Data, &p); err != nil {
+		return false, err
 	}
 
 	inMetadata, err := newMetadata(p)
 	if err != nil {
-		nackLog(msg, err)
-		return
+		return false, err
 	}
 
 	_, isReplaced := msg.Attributes["overwrittenByGeneration"]
 	eventType := msg.Attributes["eventType"]
 
+	// Ignore replacement messages
+	if isReplaced {
+		return true, nil
+	}
+
 	switch eventType {
 	case storage.ObjectFinalizeEvent:
-		if err := s.handleFinalize(inMetadata); err != nil {
-			nackLog(msg, err)
-			return
+		if err = s.handleFinalize(inMetadata); err != nil {
+			return false, err
 		}
-
 	case storage.ObjectDeleteEvent:
-		// Ignore replacement events
-		if isReplaced {
-			msg.Ack()
-			return
-		}
-
 		if err := s.handleDelete(inMetadata); err != nil {
-			nackLog(msg, err)
-			return
+			return false, err
 		}
 	case storage.ObjectArchiveEvent:
-		// Ignore replacement events
-		if isReplaced {
-			msg.Ack()
-			return
-		}
-
 		if err := s.handleArchive(inMetadata); err != nil {
-			nackLog(msg, err)
-			return
+			return false, err
 		}
 	default:
-		defaultErr := fmt.Errorf("unknown event type: %s", eventType)
-		nackLog(msg, defaultErr)
+		return false, fmt.Errorf("unknown event type: %s", eventType)
 	}
 
+	return true, nil
+}
+
+// consumeMessage is a callback function for pubsub.Receive() which handles
+// the acknowledgment of messages by calling processMessage()
+func (s *SubscriberService) consumeMessage(ctx context.Context, msg *pubsub.Message) {
+	if ack, err := processMessage(s, msg); !ack {
```

Comment:
you function so far always return `false, error` or `true, nil` so this would work. I think this function should be modified to only return error for now. In the future if it ever needs extension to other permutations it should look something like:
```
if err != nil {log; nack}
else if ack {ack}
// no action if there is no error and the child function says don't ack. The message will expire by itself.
```

Reply:
I was thinking so too. There might be a possibility in the future that we would like to Ack/Nack on some errors but it is not the case. 

Since now it is not necessary, I will apply your changes. Thank you for the insight!